### PR TITLE
fix(plasma-ui): one more refactor TimePicker

### PR DESCRIPTION
**Pull request by KateKate**
<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.102.3-canary.152.3249398287.0
  npm install @salutejs/plasma-core@1.76.3-canary.152.3249398287.0
  npm install @salutejs/plasma-icons@1.102.3-canary.152.3249398287.0
  npm install @salutejs/plasma-temple@1.102.4-canary.152.3249398287.0
  npm install @salutejs/plasma-typo@0.23.3-canary.152.3249398287.0
  npm install @salutejs/plasma-ui@1.136.3-canary.152.3249398287.0
  npm install @salutejs/plasma-web@1.134.3-canary.152.3249398287.0
  npm install @salutejs/plasma-cy-utils@0.29.3-canary.152.3249398287.0
  npm install @salutejs/plasma-sb-utils@0.75.3-canary.152.3249398287.0
  # or 
  yarn add @salutejs/plasma-b2c@1.102.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-core@1.76.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-icons@1.102.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-temple@1.102.4-canary.152.3249398287.0
  yarn add @salutejs/plasma-typo@0.23.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-ui@1.136.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-web@1.134.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-cy-utils@0.29.3-canary.152.3249398287.0
  yarn add @salutejs/plasma-sb-utils@0.75.3-canary.152.3249398287.0
  ```
</details>

:baby_chick: Download canary assets:

[sbermarket_ios-swift--canary.152.3249398287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/sbermarket_ios-swift--canary.152.3249398287.swift)  
[sbermarket_kotlin--canary.152.3249398287.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/sbermarket_kotlin--canary.152.3249398287.kt)  
[sbermarket_react-native--canary.152.3249398287.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/sbermarket_react-native--canary.152.3249398287.ts)  
[colors--canary.152.3249398287.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.152.3249398287.xml)  
[PlasmaTokensColor--canary.152.3249398287.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.152.3249398287.swift)

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

File: packages/plasma-ui/src/components/Pickers/Picker.tsx [hunk 1]
```diff
@@ -340,75 +336,91 @@ export const Picker = ({
...
-            setScrollAnim(false);
-        }
+    /**
+     * В случае изменения длины массива надо установить
+     * новый актуальный index для текущего value.
+     * Например, меняем месяц и новое кол-во дней в месяце [1 ... 30] => [1 ... 31].
+     */
+    const [prevVirtualItemsLength, setPrevVirtualItemsLength] = useState(virtualItems.length);
+    if (virtualItems.length !== prevVirtualItemsLength) {
+        unstable_batchedUpdates(() => {
```
Comment (neretin-trike) [inline packages/plasma-ui/src/components/Pickers/Picker.tsx:357]: всё-таки будем использовать этот метод? :thinking:  А есть информация, как он ведёт себя в React 18?

File: packages/plasma-ui/src/components/Pickers/Picker.tsx [hunk 2]
```diff
@@ -448,19 +459,30 @@ export const Picker = ({
...
+                return;
+            }
+
+            setIndex(i);
+        },
+        [virtualItems, infiniteScroll, max, min, isSingleItem, prevValue],
+    );
+
+    const onItemClick = useCallback(
+        (v: PickerItemType) => {
```
Comment (neretin-trike) [inline packages/plasma-ui/src/components/Pickers/Picker.tsx:479]: *минор*

я бы всё-таки не сокращал переменные до одной буквы, если можно, напиши пожалуйста value :)

File: packages/plasma-ui/src/components/Pickers/Picker.tsx [hunk 3]
```diff
@@ -340,75 +336,91 @@ export const Picker = ({
...
+         */
+        unstable_batchedUpdates(() => {
+            const firstRealItemsIndex = max + 1;
+            const lastRealItemsIndex = max * 2 + 1;
+            if (newIndex === firstRealItemsIndex || newIndex === lastRealItemsIndex) {
+                hasScrollAnimRef.current = false;
+            }
+            setIndex(newIndex);
+            setPrevValue(virtualItems[newIndex]?.value);
+        });
```
Comment (neretin-trike) [inline packages/plasma-ui/src/components/Pickers/Picker.tsx:386]: раньше в useEffect вызывался `setScrollAnim(false);`. Честно не помню для чего, но кажется это для каких-то крайних кейсов. Не нашёл в коде аналога с getDerivedStateFromProps, может тоже добавим куда-нибудь сюда `hasScrollAnimRef.current = false;`?

File: packages/plasma-ui/src/components/Pickers/Picker.tsx (hunk 1 above)
Comment (KateKate) [reply packages/plasma-ui/src/components/Pickers/Picker.tsx:357]: в React 18 бесполезен, но есть, чтобы не сломать обратную совместимость

File: packages/plasma-ui/src/components/Pickers/Picker.tsx (hunk 2 above)
Comment (KateKate) [reply packages/plasma-ui/src/components/Pickers/Picker.tsx:479]: ok

File: packages/plasma-ui/src/components/Pickers/Picker.tsx (hunk 3 above)
Comment (KateKate) [reply packages/plasma-ui/src/components/Pickers/Picker.tsx:386]: а в каком именно useEffect? кажется все перенесла как было

Comment (neretin-trike) [reply packages/plasma-ui/src/components/Pickers/Picker.tsx:386]: здесь - https://github.com/salute-developers/plasma/blob/master/packages/plasma-ui/src/components/Pickers/Picker.tsx#L410. Может всё-таки где-то есть, но я не обнаружил

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/demo-canvas-app-pr-152/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr-152/
  ui storybook: http://plasma.sberdevices.ru/pr-152/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr-152/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr-152/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr-152/temple-storybook/

**Pull request closed without merging**